lucet-runtime/lucet-runtime-internals/src/module/dl.rs

```diff
@@ -71,31 +73,106 @@ impl DlModule {
-                        "`lucet_trap_manifest` is defined but null"
+        let code_metadata_ptr = unsafe {
+            lib.get::<*const u8>(b"lucet_code_metadata").map_err(|e| {
+                lucet_incorrect_module!(
+                    "error loading required symbol `lucet_code_metadata`: {}",
+                    e
+                )
+            })?
+        };
+
+        let code_metadata_len = unsafe {
+            lib.get::<usize>(b"lucet_code_metadata_len").map_err(|e| {
+                lucet_incorrect_module!(
+                    "error loading required symbol `lucet_code_metadata_len`: {}",
+                    e
+                )
+            })?
+        };
+
+        let code_metadata_slice: &'static [u8] =
+            unsafe { slice::from_raw_parts(*code_metadata_ptr, *code_metadata_len) };
+        let mut code_metadata = CodeMetadata::deserialize(code_metadata_slice)?;
+
+        let function_manifest = unsafe {
+            let manifest_len_ptr = lib.get::<*const u32>(b"lucet_function_manifest_len");
+            let manifest_ptr = lib.get::<*const FunctionSpec>(b"lucet_function_manifest");
+
+            match (manifest_ptr, manifest_len_ptr) {
+                (Ok(ptr), Ok(len_ptr)) => {
+                    let manifest_len = len_ptr.as_ref().ok_or(lucet_incorrect_module!(
+                        "`lucet_function_manifest_len` is defined but null"
                     ))?;
-                from_raw_parts(records, *len as usize)
-            } else {
-                &[]
+                    let manifest = ptr.as_ref().ok_or(lucet_incorrect_module!(
+                        "`lucet_function_manifest` is defined but null"
+                    ))?;
+
+                    from_raw_parts(manifest, *manifest_len as usize)
+                }
+                // TODO: Can a module be data-only?
+                (Err(_), Err(_)) => &[],
+                (Ok(_), Err(e)) => {
+                    return Err(lucet_incorrect_module!(
+                        "error loading symbol `lucet_function_manifest_len`: {}",
+                        e
+                    ));
+                }
+                (Err(e), Ok(_)) => {
+                    return Err(lucet_incorrect_module!(
+                        "error loading symbol `lucet_function_manifest`: {}",
+                        e
+                    ));
+                }
             }
         };
 
+        // Now that we have a function manifest, we can fix up table_addr values in
+        // the trap manifest, which should point to the table in a readonly section
+
+        for (idx, record) in code_metadata.trap_manifest.iter_mut().enumerate() {
+            let function = function_manifest.get(record.func_index as usize).ok_or(
+                lucet_incorrect_module!(
+                    "trap manifest includes an invalid function index: {}",
+                    record.func_index,
+                ),
+            )?;
+            let fn_name = unsafe {
```

Comment:
This is super hairy, but works for the moment, so at least it's a working changeset to think about

Reply:
Thinking about it more, I think avoiding having a relocation in `TrapManifestRecord` is the wrong way to go about this - the address of the slice must be be somewhere, this just gets to it in a roundabout `dladdr` matter. Before the last commit on this branch I still had `TrapManifestRecord` written out with a relocation on `table_addr`, for comparison.

Upside is, I think, that we could get away with zero-copy deserialization of a module's `[TrapManifestRecord]` just like trap sites, and the same for other similar kinds of data (thinking unwind information, primarily)

Reply:
yep, lets avoid depending on `dladdr` in any of the new code, and also using `dlsym` (via `lib.get` for `trap_manifest_ptr`) to construct symbols from strings. Instead, I'd change `FunctionSpec` to have the fields `pub code: *const c_void, pub traps: *const c_void, len: u32` and, when serializing, use relocations for the first two members to point to the code and trap symbols.